File: python/src/unround/colour.py
# SPDX-FileCopyrightText: 2026 Yuma Kakei <yumasansansan@gmail.com>
# SPDX-License-Identifier: GPL-3.0-or-later
"""The YCbCr of JFIF and its inverse, in binary64 (docs/math.md, 8).

The coefficients are the rationals that K_R = 0.299 and K_B = 0.114 make, each the
double nearest to it, and not the six-digit decimals that T.871 prints: those would
move a converted block's coefficients by more than 10^-4 of a small step. Each
operation is rounded to the nearest in the order docs/math.md gives, none fused, so
that the RGB of a solution is the same to the last bit wherever it is computed.
"""

from fractions import Fraction
from typing import Final

import numpy as np
import numpy.typing as npt

__all__ = ["to_rgb", "to_ycbcr"]

type Array = npt.NDArray[np.float64]

_CENTRE: Final = 128.0
_CHANNELS: Final = 3

_RED_FROM_CR: Final = float(Fraction(701, 500))  # 2 (1 - K_R) = 1.402
_BLUE_FROM_CB: Final = float(Fraction(443, 250))  # 2 (1 - K_B) = 1.772
_GREEN_FROM_CB: Final = float(Fraction(25251, 73375))  # 2 K_B (1 - K_B) / K_G
_GREEN_FROM_CR: Final = float(Fraction(209599, 293500))  # 2 K_R (1 - K_R) / K_G

_Y_FROM_RED: Final = float(Fraction(299, 1000))
_Y_FROM_GREEN: Final = float(Fraction(587, 1000))
_Y_FROM_BLUE: Final = float(Fraction(114, 1000))
_CB_FROM_BLUE: Final = float(Fraction(250, 443))  # 1 / 1.772
_CR_FROM_RED: Final = float(Fraction(500, 701))  # 1 / 1.402


def to_rgb(planes: npt.ArrayLike) -> Array:
    """The RGB of JFIF, (height, width, 3), of Y, Cb and Cr planes, (3, height, width), in binary64.

    R = Y + c_R (Cr - 128), B = Y + c_B (Cb - 128), G = (Y - c_GB (Cb - 128)) - c_GR (Cr - 128).
    """
    luma, blue_difference, red_difference = _planes(planes)
    cb = blue_difference - _CENTRE
    cr = red_difference - _CENTRE
    red = luma + _RED_FROM_CR * cr
    blue = luma + _BLUE_FROM_CB * cb
    green = (luma - _GREEN_FROM_CB * cb) - _GREEN_FROM_CR * cr
    return np.stack((red, green, blue), axis=-1)


def to_ycbcr(picture: npt.ArrayLike) -> Array:
    """The Y, Cb and Cr planes of JFIF, (3, height, width), of an RGB picture, (height, width, 3), in binary64.

    Y = (k_R R + k_G G) + k_B B, Cb = 128 + k_Cb (B - Y), Cr = 128 + k_Cr (R - Y).
    """
    samples = np.asarray(picture, dtype=np.float64)
    if samples.ndim != _CHANNELS or samples.shape[2] != _CHANNELS:
        message = f"an RGB picture is (height, width, 3), not {samples.shape}"
        raise ValueError(message)
    red, green, blue = samples[:, :, 0], samples[:, :, 1], samples[:, :, 2]
    luma = (_Y_FROM_RED * red + _Y_FROM_GREEN * green) + _Y_FROM_BLUE * blue
    blue_difference = _CENTRE + _CB_FROM_BLUE * (blue - luma)
    red_difference = _CENTRE + _CR_FROM_RED * (red - luma)
    return np.stack((luma, blue_difference, red_difference))


def _planes(planes: npt.ArrayLike) -> tuple[Array, Array, Array]:
    samples = np.asarray(planes, dtype=np.float64)
    if samples.ndim != _CHANNELS or samples.shape[0] != _CHANNELS:
        message = f"YCbCr planes are (3, height, width), not {samples.shape}"
        raise ValueError(message)
    return samples[0], samples[1], samples[2]

File: rust/unround/src/colour.rs
// SPDX-FileCopyrightText: 2026 Yuma Kakei <yumasansansan@gmail.com>
// SPDX-License-Identifier: GPL-3.0-or-later

//! The YCbCr of JFIF and its inverse, in binary64 (docs/math.md, 8).
//!
//! The coefficients are the rationals that `K_R = 0.299` and `K_B = 0.114` make,
//! each the double nearest to it (the quotient of two integers that binary64 holds
//! exactly, which one division rounds correctly), and not the six-digit decimals
//! that T.871 prints: those would move a converted block's coefficients by more
//! than 1e-4 of a small step. The operations are in the order docs/math.md gives;
//! each product is fused with the sum that takes it where the processor can.

use crate::dct::multiply_add;
use crate::error::Error;

const CENTRE: f64 = 128.0;

/// `2 (1 - K_R) = 1.402 = 701 / 500`.
pub const RED_FROM_CR: f64 = 701.0 / 500.0;
/// `2 (1 - K_B) = 1.772 = 443 / 250`.
pub const BLUE_FROM_CB: f64 = 443.0 / 250.0;
/// `2 K_B (1 - K_B) / K_G = 25251 / 73375`.
pub const GREEN_FROM_CB: f64 = 25_251.0 / 73_375.0;
/// `2 K_R (1 - K_R) / K_G = 209599 / 293500`.
pub const GREEN_FROM_CR: f64 = 209_599.0 / 293_500.0;
/// `K_R = 299 / 1000`.
pub const Y_FROM_RED: f64 = 299.0 / 1000.0;
/// `K_G = 587 / 1000`.
pub const Y_FROM_GREEN: f64 = 587.0 / 1000.0;
/// `K_B = 114 / 1000`.
pub const Y_FROM_BLUE: f64 = 114.0 / 1000.0;
/// `1 / 1.772 = 250 / 443`.
pub const CB_FROM_BLUE: f64 = 250.0 / 443.0;
/// `1 / 1.402 = 500 / 701`.
pub const CR_FROM_RED: f64 = 500.0 / 701.0;

/// The RGB of JFIF of one pixel's Y, Cb and Cr:
/// `R = Y + c_R (Cr - 128)`, `B = Y + c_B (Cb - 128)`,
/// `G = (Y - c_GB (Cb - 128)) - c_GR (Cr - 128)`.
#[inline]
#[must_use]
pub fn rgb_of(luma: f64, blue_difference: f64, red_difference: f64) -> [f64; 3] {
    let cb = blue_difference - CENTRE;
    let cr = red_difference - CENTRE;
    let red = multiply_add(RED_FROM_CR, cr, luma);
    let blue = multiply_add(BLUE_FROM_CB, cb, luma);
    let green = multiply_add(-GREEN_FROM_CR, cr, multiply_add(-GREEN_FROM_CB, cb, luma));
    [red, green, blue]
}

/// The Y, Cb and Cr of JFIF of one RGB pixel: `Y = (k_R R + k_G G) + k_B B`,
/// `Cb = 128 + k_Cb (B - Y)`, `Cr = 128 + k_Cr (R - Y)`.
#[inline]
#[must_use]
pub fn ycbcr_of(red: f64, green: f64, blue: f64) -> [f64; 3] {
    let luma = multiply_add(Y_FROM_BLUE, blue, multiply_add(Y_FROM_RED, red, Y_FROM_GREEN * green));
    let blue_difference = multiply_add(CB_FROM_BLUE, blue - luma, CENTRE);
    let red_difference = multiply_add(CR_FROM_RED, red - luma, CENTRE);
    [luma, blue_difference, red_difference]
}

/// The RGB of JFIF, `height x width x 3` interleaved, of Y, Cb and Cr planes of
/// `height x width` each, one after another.
///
/// # Errors
///
/// [`Error::Options`] unless there are three planes of that size.
pub fn to_rgb(planes: &[f64], height: usize, width: usize) -> Result<Vec<f64>, Error> {
    let size = height * width;
    if planes.len() != 3 * size {
        return Err(Error::Options(format!(
            "Y, Cb and Cr planes of {height} x {width} are {} samples, not {}",
            3 * size,
            planes.len()
        )));
    }
    let (luma, rest) = planes.split_at(size);
    let (blue, red) = rest.split_at(size);
    let mut picture = Vec::with_capacity(3 * size);
    for ((&y, &cb), &cr) in luma.iter().zip(blue).zip(red) {
        picture.extend_from_slice(&rgb_of(y, cb, cr));
    }
    Ok(picture)
}

/// The Y, Cb and Cr planes of JFIF, one after another, of an RGB picture of
/// `height x width x 3` interleaved.
///
/// # Errors
///
/// [`Error::Options`] unless the picture has that size.
pub fn to_ycbcr(picture: &[f64], height: usize, width: usize) -> Result<Vec<f64>, Error> {
    let size = height * width;
    if picture.len() != 3 * size {
        return Err(Error::Options(format!(
            "an RGB picture of {height} x {width} is {} samples, not {}",
            3 * size,
            picture.len()
        )));
    }
    let mut planes = vec![0.0; 3 * size];
    for (index, &[red, green, blue]) in picture.as_chunks::<3>().0.iter().enumerate() {
        let [y, cb, cr] = ycbcr_of(red, green, blue);
        planes[index] = y;
        planes[size + index] = cb;
        planes[2 * size + index] = cr;
    }
    Ok(planes)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::exact::quotient;

    #[test]
    #[expect(clippy::float_cmp, reason = "constants are compared to the last bit")]
    fn the_constants_are_the_nearest_doubles_to_their_rationals() {
        assert_eq!(RED_FROM_CR, quotient(701, 500));
        assert_eq!(BLUE_FROM_CB, quotient(443, 250));
        assert_eq!(GREEN_FROM_CB, quotient(25_251, 73_375));
        assert_eq!(GREEN_FROM_CR, quotient(209_599, 293_500));
        assert_eq!(Y_FROM_RED, 0.299);
        assert_eq!(Y_FROM_GREEN, 0.587);
        assert_eq!(Y_FROM_BLUE, 0.114);
        assert_eq!(CB_FROM_BLUE, quotient(250, 443));
        assert_eq!(CR_FROM_RED, quotient(500, 701));
    }

    #[test]
    #[expect(clippy::float_cmp, reason = "a grey converts exactly")]
    fn a_grey_is_its_luma() {
        // With Cb and Cr at 128, every product is 0 and R = G = B = Y exactly.
        for luma in [0.0, 17.25, 128.0, 254.5, -3.0, 300.0] {
            assert_eq!(rgb_of(luma, 128.0, 128.0), [luma; 3]);
        }
    }
}
